Decode panic when handling bad input

**Describe the bug**
When attempting to parse a table that is expected to be an array/slice, unmarshaling panics.

**To Reproduce**

```go
func Test_Decode(t *testing.T) {
	definition := struct {
		Things []string
	}{}

	values := `[things]
foo = "bar"`

	if err := toml.Unmarshal([]byte(values), &definition); err != nil {
		t.Fail()
	}
}

```


**Expected behavior**

Ideally the parser should return a `DecodeError` rather than panic.

**Versions**
 - go-toml: `v2.0.0-beta.3`
 - go: `go version go1.17 darwin/arm64`
 - operating system: `macOS`

**Additional context**

Panic trace, appears due to unsafe access in `(reflect.Value).Index`

```
|| --- FAIL: Test_Decode (0.00s)
|| 
|| goroutine 4 [running]:
|| testing.tRunner.func1.2({0x104b58e20, 0x104bff748})
|| 	/opt/homebrew/Cellar/go/1.17/libexec/src/testing/testing.go:1209 +0x258
|| testing.tRunner.func1(0x14000125520)
|| 	/opt/homebrew/Cellar/go/1.17/libexec/src/testing/testing.go:1212 +0x284
|| panic({0x104b58e20, 0x104bff748})
|| 	/opt/homebrew/Cellar/go/1.17/libexec/src/runtime/panic.go:1038 +0x21c
|| reflect.Value.Index({0x104b51a20, 0x1400000e420, 0x197}, 0xffffffffffffffff)
|| 	/opt/homebrew/Cellar/go/1.17/libexec/src/reflect/value.go:1297 +0x1e0
|| github.com/pelletier/go-toml/v2.(*decoder).handleTable(0x1400011bea0, {0x1, 0x140002fe268}, {0x104b51a20, 0x1400000e420, 0x197})
|| 	/Users/christiangregg/go/pkg/mod/github.com/pelletier/go-toml/v2@v2.0.0-beta.3/unmarshaler.go:472 +0x16c
|| github.com/pelletier/go-toml/v2.(*decoder).handleKeyPart(0x1400011bea0, {0x1, 0x140002fe268}, {0x104b83aa0, 0x1400000e420, 0x199}, 0x1400005cc78, 0x104bfceb8)
|| 	/Users/christiangregg/go/pkg/mod/github.com/pelletier/go-toml/v2@v2.0.0-beta.3/unmarshaler.go:426 +0xcc
|| github.com/pelletier/go-toml/v2.(*decoder).handleTablePart(0x1400005cea0, {0x1, 0x140002fe268}, {0x104b83aa0, 0x1400000e420, 0x199})
|| 	/Users/christiangregg/go/pkg/mod/github.com/pelletier/go-toml/v2@v2.0.0-beta.3/unmarshaler.go:532 +0x78
|| github.com/pelletier/go-toml/v2.(*decoder).handleTable(0x1400005cea0, {0x1, 0x140002fe268}, {0x104b83aa0, 0x1400000e420, 0x199})
|| 	/Users/christiangregg/go/pkg/mod/github.com/pelletier/go-toml/v2@v2.0.0-beta.3/unmarshaler.go:484 +0xa4
|| github.com/pelletier/go-toml/v2.(*decoder).handleRootExpression(0x1400011bea0, 0x140002fe230, {0x104b83aa0, 0x1400000e420, 0x199})
|| 	/Users/christiangregg/go/pkg/mod/github.com/pelletier/go-toml/v2@v2.0.0-beta.3/unmarshaler.go:230 +0x178
|| github.com/pelletier/go-toml/v2.(*decoder).fromParser(0x1400011bea0, {0x104b83aa0, 0x1400000e420, 0x199})
|| 	/Users/christiangregg/go/pkg/mod/github.com/pelletier/go-toml/v2@v2.0.0-beta.3/unmarshaler.go:192 +0xb0
|| github.com/pelletier/go-toml/v2.(*decoder).FromParser(0x1400011bea0, {0x104b4bda0, 0x1400000e420})
|| 	/Users/christiangregg/go/pkg/mod/github.com/pelletier/go-toml/v2@v2.0.0-beta.3/unmarshaler.go:177 +0x140
|| github.com/pelletier/go-toml/v2.Unmarshal({0x14000026498, 0x14, 0x18}, {0x104b4bda0, 0x1400000e420})
|| 	/Users/christiangregg/go/pkg/mod/github.com/pelletier/go-toml/v2@v2.0.0-beta.3/unmarshaler.go:27 +0xf4

```

---

Happy to take a crack at fixing this issue, if it can be confirmed as unexpected behaviour! 😄

--- a/unmarshaler.go
+++ b/unmarshaler.go
@@ -488,6 +488,9 @@ func (d *decoder) handleArrayTablePart(key ast.Iterator, v reflect.Value) (refle
 // cannot handle it.
 func (d *decoder) handleTable(key ast.Iterator, v reflect.Value) (reflect.Value, error) {
 	if v.Kind() == reflect.Slice {
+		if v.Len() == 0 {
+			return reflect.Value{}, newDecodeError(key.Node().Data, "cannot store a table in a slice")
+		}
 		elem := v.Index(v.Len() - 1)
 		x, err := d.handleTable(key, elem)
 		if err != nil {

--- a/unmarshaler_test.go
+++ b/unmarshaler_test.go
@@ -1810,6 +1810,18 @@ func TestIssue602(t *testing.T) {
 	require.Equal(t, expected, v)
 }
 
+func TestIssue623(t *testing.T) {
+	definition := struct {
+		Things []string
+	}{}
+
+	values := `[things]
+foo = "bar"`
+
+	err := toml.Unmarshal([]byte(values), &definition)
+	require.Error(t, err)
+}
+
 //nolint:funlen
 func TestUnmarshalDecodeErrors(t *testing.T) {
 	examples := []struct {

